File: backend/smurfing_detector.py
"""
Smurfing Detector (Fan-in / Fan-out)
=====================================

Detects:
1. Fan-in: Multiple accounts send to one aggregator (10+ senders -> 1 receiver)
2. Fan-out: One account disperses to many receivers (1 sender -> 10+ receivers)
3. Temporal analysis: Transactions within a 72-hour window

"""

from collections import defaultdict
from datetime import datetime, timedelta

# ==============================
# CONFIG
# ==============================

FAN_IN_THRESHOLD = 10
FAN_OUT_THRESHOLD = 10
TIME_WINDOW_HOURS = 72

class SmurfingDetector:

    def __init__(self):
        self.fan_in = defaultdict(list)
        self.fan_out = defaultdict(list)

    def parse_ts(self, ts):
        if not ts:
            return None
        for fmt in ("%Y-%m-%d %H:%M:%S", "%d-%m-%Y %H:%M:%S", "%d-%m-%Y %H:%M", "%Y-%m-%d"):
            try:
                return datetime.strptime(ts.strip(), fmt)
            except ValueError:
                continue
        return None

    def run(self, transactions):
        self.fan_in.clear()
        self.fan_out.clear()

        # Group transactions by receiver (fan-in) and sender (fan-out)
        for txn in transactions:
            s = txn.get("sender_id")
            r = txn.get("receiver_id")
            ts = self.parse_ts(txn.get("timestamp"))
            amount = float(txn.get("amount", 0))

            if s and r and ts:
                self.fan_in[r].append({"sender": s, "ts": ts, "amount": amount})
                self.fan_out[s].append({"receiver": r, "ts": ts, "amount": amount})

        suspicious_accounts = []
        rings = []

        # Detect Fan-in
        for acc, txns in self.fan_in.items():
            # Sort by timestamp
            txns.sort(key=lambda x: x["ts"])
            
            # Slater window check (simplified for now: checking if ANY 72h window has > threshold unique senders)
            # A more robust sliding window could be implemented, but this is a good start.
            
            # Correct approach: Slide a window over the sorted transactions
            unique_senders_in_window = set()
            start_idx = 0
            
            for i in range(len(txns)):
                current_ts = txns[i]["ts"]
                
                # Advance start_idx to keep window within 72 hours
                while start_idx < i and (current_ts - txns[start_idx]["ts"]) > timedelta(hours=TIME_WINDOW_HOURS):
                    start_idx += 1
                
                # Check unique senders in the current window [start_idx, i]
                current_window = txns[start_idx : i+1]
                senders = set(t["sender"] for t in current_window)
                
                if len(senders) >= FAN_IN_THRESHOLD:
                    suspicious_accounts.append({
                        "account_id": acc,
                        "suspicion_score": 80.0, # Base high score for smurfing
                        "detected_patterns": ["fan_in"],
                        "ring_id": f"SMURF_IN_{acc}"
                    })
                    
                    rings.append({
                        "ring_id": f"SMURF_IN_{acc}",
                        "member_accounts": list(senders) + [acc],
                        "pattern_type": "fan_in",
                        "risk_score": 85.0,
                        "detected_at": current_ts.strftime("%Y-%m-%d %H:%M:%S")
                    })
                    break # Flag once per account

        # Detect Fan-out
        for acc, txns in self.fan_out.items():
            txns.sort(key=lambda x: x["ts"])
            
            start_idx = 0
            for i in range(len(txns)):
                current_ts = txns[i]["ts"]
                
                while start_idx < i and (current_ts - txns[start_idx]["ts"]) > timedelta(hours=TIME_WINDOW_HOURS):
                    start_idx += 1
                
                current_window = txns[start_idx : i+1]
                receivers = set(t["receiver"] for t in current_window)
                
                if len(receivers) >= FAN_OUT_THRESHOLD:
                    suspicious_accounts.append({
                        "account_id": acc,
                        "suspicion_score": 80.0,
                        "detected_patterns": ["fan_out"],
                        "ring_id": f"SMURF_OUT_{acc}"
                    })
                    
                    rings.append({
                        "ring_id": f"SMURF_OUT_{acc}",
                        "member_accounts": [acc] + list(receivers),
                        "pattern_type": "fan_out",
                        "risk_score": 85.0,
                        "detected_at": current_ts.strftime("%Y-%m-%d %H:%M:%S")
                    })
                    break

        return {
            "suspicious_accounts": suspicious_accounts,
            "fraud_rings": rings
        }

File: src/lib/detection/SmurfingDetector.js
/**
 * Smurfing Detector
 * 
 * Detects:
 * 1. Fan-in: Multiple accounts -> One aggregator (10+ senders)
 * 2. Fan-out: One aggregator -> Multiple receivers (1 sender -> 10+ receivers)
 * 3. Temporal Analysis: 72-hour window.
 */

export class SmurfingDetector {
    constructor(transactions) {
        this.transactions = transactions;
        this.fanIn = new Map(); // receiver -> [{sender, ts, amount}]
        this.fanOut = new Map(); // sender -> [{receiver, ts, amount}]
    }

    run() {
        this.transactions.forEach(txn => {
            const u = txn.sender_id;
            const v = txn.receiver_id;
            const amount = parseFloat(txn.amount);
            const ts = new Date(txn.timestamp);

            if (!this.fanIn.has(v)) this.fanIn.set(v, []);
            this.fanIn.get(v).push({ sender: u, ts, amount });

            if (!this.fanOut.has(u)) this.fanOut.set(u, []);
            this.fanOut.get(u).push({ receiver: v, ts, amount });
        });

        const suspiciousAccounts = [];
        const rings = [];
        const THRESHOLD = 10;
        const TIME_WINDOW_MS = 72 * 60 * 60 * 1000;

        // Detect Fan-In
        this.fanIn.forEach((txns, receiver) => {
            txns.sort((a, b) => a.ts - b.ts);

            for (let i = 0; i < txns.length; i++) {
                let startIdx = i;
                const windowEnd = txns[i].ts;
                const windowStart = new Date(windowEnd.getTime() - TIME_WINDOW_MS);

                // Find all txns in [windowStart, windowEnd]
                // Since we iterate i forward, we look back or we assume [start, i] is the window.
                // Let's use a sliding window approach properly.
            }

            // Simpler: iterate all windows
            let left = 0;
            const uniqueSenders = new Map(); // sender -> count in window

            for (let right = 0; right < txns.length; right++) {
                const currentTxn = txns[right];
                uniqueSenders.set(currentTxn.sender, (uniqueSenders.get(currentTxn.sender) || 0) + 1);

                while (currentTxn.ts - txns[left].ts > TIME_WINDOW_MS) {
                    const leftTxn = txns[left];
                    const count = uniqueSenders.get(leftTxn.sender);
                    if (count === 1) uniqueSenders.delete(leftTxn.sender);
                    else uniqueSenders.set(leftTxn.sender, count - 1);
                    left++;
                }

                if (uniqueSenders.size >= THRESHOLD) {
                    // Found Fan-in
                    const ringId = `SMURF_IN_${receiver}`;
                    rings.push({
                        ring_id: ringId,
                        member_accounts: [receiver, ...uniqueSenders.keys()],
                        pattern_type: "fan_in",
                        risk_score: 85.0, // High risk
                        detected_at: currentTxn.ts.toISOString().replace('T', ' ').split('.')[0]
                    });

                    suspiciousAccounts.push({
                        account_id: receiver,
                        suspicion_score: 85.0,
                        detected_patterns: ["fan_in"],
                        ring_id: ringId
                    });

                    // Avoid duplicates for same receiver? 
                    // We might want just one alert per receiver
                    return; // optimize: break outer loop for this receiver
                }
            }
        });

        // Detect Fan-Out
        this.fanOut.forEach((txns, sender) => {
            txns.sort((a, b) => a.ts - b.ts);

            let left = 0;
            const uniqueReceivers = new Map();

            for (let right = 0; right < txns.length; right++) {
                const currentTxn = txns[right];
                uniqueReceivers.set(currentTxn.receiver, (uniqueReceivers.get(currentTxn.receiver) || 0) + 1);

                while (currentTxn.ts - txns[left].ts > TIME_WINDOW_MS) {
                    const leftTxn = txns[left];
                    const count = uniqueReceivers.get(leftTxn.receiver);
                    if (count === 1) uniqueReceivers.delete(leftTxn.receiver);
                    else uniqueReceivers.set(leftTxn.receiver, count - 1);
                    left++;
                }

                if (uniqueReceivers.size >= THRESHOLD) {
                    const ringId = `SMURF_OUT_${sender}`;
                    rings.push({
                        ring_id: ringId,
                        member_accounts: [sender, ...uniqueReceivers.keys()],
                        pattern_type: "fan_out",
                        risk_score: 85.0,
                        detected_at: currentTxn.ts.toISOString().replace('T', ' ').split('.')[0]
                    });

                    suspiciousAccounts.push({
                        account_id: sender,
                        suspicion_score: 85.0,
                        detected_patterns: ["fan_out"],
                        ring_id: ringId
                    });
                    return;
                }
            }
        });

        return { rings, suspiciousAccounts };
    }
}
